File: web/sub_free.py
'''
免费节点获取服务

poetry run hypercorn sub_free:app --bind 127.0.0.1:5333 --workers
poetry run hypercorn sub_free:app --bind 0.0.0.0:5333 --workers 9
poetry run hypercorn sub_free:app --bind [::]:5333 --workers 9
poetry run gunicorn sub_free:app -b 0.0.0.0:5333 -w 9

1Panel: deploy on Python Environment
run cmd: pip install poetry && poetry install && poetry run hypercorn sub_free:app --bind 0.0.0.0:5333 --workers 9


'''
from flask import Flask, request, jsonify, send_from_directory, request, abort
from pathlib import Path

import os
import requests

from sub_url import * 


app = Flask(__name__)

HOST  = "localhost"
PORT  = 5003
DEBUG = False
DEBUG = True
FLD_DATA = 'data'

# 指定文件根目录
BASE_DIRECTORY = Path(FLD_DATA).resolve()

# 检查文件根目录是否存在，如果不存在则创建
BASE_DIRECTORY.mkdir(parents=True, exist_ok=True)

fpath = f'{FLD_DATA}/test.txt'
if not os.path.exists(fpath): 
    with open(fpath, 'w') as file:
        file.write(f' This is a file for remote url request test. ')

def download_file_to_local(url, filename):
    '''下载链接中的文件，并将其保存为filename定义的文件名'''
    response = requests.get(url)
    if response.status_code == 200:
        with open(filename, "wb") as file:
            file.write(response.content)
        # print("文件已成功下载并保存为", filename)
    else:
        print("无法访问链接:", url)        
        
@app.route('/download/<path:filepath>', methods=['GET'])
def route_download_file(filepath):
    # 将请求路径拼接到根目录，确保路径合法
    file_path = (BASE_DIRECTORY / filepath).resolve()

    # 确保文件路径在根目录内，防止目录遍历攻击
    if not file_path.is_file() or BASE_DIRECTORY not in file_path.parents:
        abort(404)  # 文件未找到则返回404错误

    # 返回文件下载
    return send_from_directory(BASE_DIRECTORY, filepath, as_attachment=True)

# 文件上传（可选），允许将文件上传到服务器的子目录
@app.route('/upload/<path:filepath>', methods=['POST'])
def route_upload_file(filepath):
    if 'file' not in request.files:
        return "No file part", 400

    file = request.files['file']
    if file.filename == '':
        return "No selected file", 400

    # 将上传的文件保存到指定的子目录
    upload_path = (BASE_DIRECTORY / filepath).resolve()
    if BASE_DIRECTORY not in upload_path.parents:
        abort(403)  # 禁止访问超出根目录的路径

    # 创建子目录（如果不存在）
    upload_path.parent.mkdir(parents=True, exist_ok=True)

    # 保存文件
    file.save(upload_path)
    return "File uploaded successfully", 200


@app.route('/c/mibei', methods=['GET'])
def route_clash_mibei():
    
    urlv,urlc = get_sub_mibei(get_urls_mibei()[0])
    
    response = requests.get(urlc)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/v/mibei', methods=['GET'])
def route_v2ray_mibei():
    
    urlv,urlc = get_sub_mibei(get_urls_mibei()[0])
    
    response = requests.get(urlv)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/c/v2rayshare', methods=['GET'])
def route_clash_v2rayshare():
    
    urlv,urlc = get_sub_v2rayshare(get_urls_v2rayshare()[0])
    
    response = requests.get(urlc)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/v/v2rayshare', methods=['GET'])
def route_v2ray_v2rayshare():
    
    urlv,urlc = get_sub_v2rayshare(get_urls_v2rayshare()[0])
    
    response = requests.get(urlv)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/c/openrunner', methods=['GET'])
def route_clash_openrunner():
    
    urlv,urlc = get_sub_openrunner(get_urls_openrunner()[0])
    
    response = requests.get(urlc)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/v/openrunner', methods=['GET'])
def route_v2ray_openrunner():
    
    urlv,urlc = get_sub_openrunner(get_urls_openrunner()[0])
    
    response = requests.get(urlv)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/c/nodefree', methods=['GET'])
def route_clash_nodefree():
    
    urlv,urlc = get_sub_nodefree(get_urls_nodefree()[0])
    
    response = requests.get(urlc)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/v/nodefree', methods=['GET'])
def route_v2ray_nodefree():
    
    urlv,urlc = get_sub_nodefree(get_urls_nodefree()[0])
    
    response = requests.get(urlv)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/c/clashnode', methods=['GET'])
def route_clash_clashnode(): 
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
        }
    
    urlv,urlc,urls = get_sub_clashnode(get_urls_clashnode()[0])
    
    response = requests.get(urlc[0],headers=headers)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/v/clashnode', methods=['GET'])
def route_v2ray_clashnode():    
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
        }
    urlv,urlc,urls = get_sub_clashnode(get_urls_clashnode()[0])
    
    url = urlv[0]
    response = requests.get(url,headers=headers)
    if DEBUG: print(f'url: {url}')
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/s/clashnode', methods=['GET'])
def route_singbox_clashnode(): 
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
        }
    
    urlv,urlc,urls = get_sub_clashnode(get_urls_clashnode()[0])
    
    response = requests.get(urls[0],headers=headers)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/b/changfeng', methods=['GET'])
def route_base64_changfeng():
    
    bs64,urlv,urlc,urlm,urls = get_sub_changfeng(get_urls_changfeng()[0])
    
    return bs64


@app.route('/c/changfeng', methods=['GET'])
def route_clash_changfeng():
    
    bs64,urlv,urlc,urlm,urls = get_sub_changfeng(get_urls_changfeng()[0])
    
    response = requests.get(urlc)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/v/changfeng', methods=['GET'])
def route_v2ray_changfeng():
    
    bs64,urlv,urlc,urlm,urls = get_sub_changfeng(get_urls_changfeng()[0])
    
    response = requests.get(urlv)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/s/changfeng', methods=['GET'])
def route_singbox_changfeng():
    
    bs64,urlv,urlc,urlm,urls = get_sub_changfeng(get_urls_changfeng()[0])
    
    response = requests.get(urls)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None

@app.route('/m/changfeng', methods=['GET'])
def route_mihomo_changfeng():
    
    bs64,urlv,urlc,urlm,urls = get_sub_changfeng(get_urls_changfeng()[0])
    
    response = requests.get(urlm)
    
    if response.status_code == 200:
        # return response.content
        return response.text
    else: 
        return None
    
@app.route('/usage', methods=['GET'])
def route_usage():
    # host_url = str(request.base_url)
    host_url = str(request.host_url)
    
    html_content = f"""
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Links Usage List</title>
</head>
<body>
    <h1>Usage Links</h1>
    <h2>米贝</h2>
    <a href="https://www.mibei77.com"/>
    <ul>
        <li><a href="{host_url}c/mibei" target="_blank">{host_url}c/mibei</a></li>\n
        <li><a href="{host_url}v/mibei" target="_blank">{host_url}v/mibei</a></li>\n\n
    </ul>
    <h2>NodeFree</h2>
    <ul>
        <li><a href="{host_url}c/nodefree" target="_blank">{host_url}c/nodefree</a></li>\n
        <li><a href="{host_url}v/nodefree" target="_blank">{host_url}v/nodefree</a></li>\n\n
    </ul>
    <h2>V2rayFree</h2>
    <ul>
        <li><a href="{host_url}c/v2rayshare" target="_blank">{host_url}c/v2rayshare</a></li>\n
        <li><a href="{host_url}v/v2rayshare" target="_blank">{host_url}v/v2rayshare</a></li>\n\n
    </ul>
    <h2>OpenRunner</h2>
    <ul>
        <li><a href="{host_url}c/openrunner" target="_blank">{host_url}c/openrunner</a></li>\n
        <li><a href="{host_url}v/openrunner" target="_blank">{host_url}v/openrunner</a></li>\n\n
    </ul>
    <h2>clashNode</h2>
    <ul>
        <li><a href="{host_url}c/clashnode" target="_blank">{host_url}c/clashnode</a></li>\n
        <li><a href="{host_url}v/clashnode" target="_blank">{host_url}v/clashnode</a></li>\n
        <li><a href="{host_url}s/clashnode" target="_blank">{host_url}s/clashnode</a></li>\n\n
    </ul>
    <h2>长风</h2>
    <ul>
        <li><a href="{host_url}c/changfeng" target="_blank">{host_url}c/changfeng</a></li>\n
        <li><a href="{host_url}v/changfeng" target="_blank">{host_url}v/changfeng</a></li>\n
        <li><a href="{host_url}s/changfeng" target="_blank">{host_url}s/changfeng</a></li>\n
        <li><a href="{host_url}m/changfeng" target="_blank">{host_url}m/changfeng</a></li>\n
    </ul>
</body>
</html>
"""
    sss = f'''
    \n
    \n{host_url}c/mibei
    \n{host_url}v/mibei
    \n
    \n{host_url}c/nodefree
    \n{host_url}v/nodefree
    \n
    \n{host_url}c/v2rayshare
    \n{host_url}v/v2rayshare
    \n
    \n{host_url}c/openrunner
    \n{host_url}v/openrunner
    \n
    \n{host_url}c/clashnode
    \n{host_url}v/clashnode
    \n{host_url}s/clashnode
    \n
    \n{host_url}c/changfeng
    \n{host_url}v/changfeng
    \n{host_url}s/changfeng
    \n{host_url}m/changfeng

    '''
    
    return html_content 
    
def usage():
    print(f'\n')
    print(f' http://{host_url}/c/mibei')
    print(f' http://{host_url}/v/mibei')
    
    # print(f'\n')
    print(f' http://{host_url}/c/nodefree')
    print(f' http://{host_url}/v/nodefree')
    
    # print(f'\n')
    print(f' http://{host_url}/c/v2rayshare')
    print(f' http://{host_url}/v/v2rayshare')
    
    # print(f'\n')
    print(f' http://{host_url}/c/openrunner')
    print(f' http://{host_url}/v/openrunner')
    
    # print(f'\n')
    print(f' http://{host_url}/c/clashnode')
    print(f' http://{host_url}/v/clashnode')
    print(f' http://{host_url}/s/clashnode')
    
    # print(f'\n')
    print(f' http://{host_url}/c/changfeng')
    print(f' http://{host_url}/v/changfeng')
    print(f' http://{host_url}/s/changfeng')
    print(f' http://{host_url}/m/changfeng')
    
    print(f'\n')

#==========================
if __name__ == '__main__':
    usage()
    app.run(host=HOST,port=PORT,debug=DEBUG)
    
    

File: web/sub_free.js
// ==========  工具函数  ==========
const USER_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36';

// 统一 fetch 包装：自动带 UA、10 s 超时
async function fetchText(url, opts = {}) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 10000);
  try {
    const res = await fetch(url, {
      headers: { 'User-Agent': USER_AGENT },
      signal: controller.signal,
      ...opts,
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    return await res.text();
  } finally {
    clearTimeout(timer);
  }
}

// 简单 HTML 解析：返回第一个匹配属性的数组
function parseLinks(html, selector, attr = 'href') {
  const regex = new RegExp(`${attr}=["']([^"']+)["']`, 'g');
  const out = [];
  let m;
  while ((m = regex.exec(html)) !== null) {
    if (m[1].startsWith('http')) out.push(m[1]);
  }
  return out;
}

// ==========  各站抓取逻辑  ==========
const SOURCES = {
  // 米贝
  async mibei() {
    const root = await fetchText('https://www.mibei77.com');
    const links = parseLinks(root, 'article a');
    const latest = links[0];
    const page = await fetchText(latest);
    const txt = parseLinks(page, 'p', 'href').find(u => u.endsWith('.txt')) || '';
    const yaml = parseLinks(page, 'p', 'href').find(u => u.endsWith('.yaml')) || '';
    return { v2ray: txt, clash: yaml };
  },

  // 长风
  async changfeng() {
    const root = await fetchText('https://www.cfmem.com/');
    const links = parseLinks(root, 'h2 a');
    const latest = links[0];
    const page = await fetchText(latest);
    const base64 = (page.match(/节点Base64.*?<pre[^>]*>([^<]+)<\/pre>/is) || [])[1]?.trim() || '';
    const subs = parseLinks(page, 'div', 'href');
    return {
      base64,
      v2ray: subs[0] || '',
      clash: subs[1] || '',
      mihomo: subs[2] || '',
      singbox: subs[3] || '',
    };
  },

  // 以下站点结构类似，只给出最简实现，可按需补全
  async nodefree() {
    const root = await fetchText('https://nodefree.org');
    const links = parseLinks(root, 'h2 a');
    const page = await fetchText(links[0]);
    const txt = parseLinks(page, 'p', 'href').find(u => u.endsWith('.txt')) || '';
    const yaml = parseLinks(page, 'p', 'href').find(u => u.endsWith('.yaml')) || '';
    return { v2ray: txt, clash: yaml };
  },

  async v2rayshare() {
    const root = await fetchText('https://v2rayshare.com/');
    const links = parseLinks(root, 'div.list-body a');
    const page = await fetchText(links[0]);
    const txt = parseLinks(page, 'p', 'href').find(u => u.endsWith('.txt')) || '';
    const yaml = parseLinks(page, 'p', 'href').find(u => u.endsWith('.yaml')) || '';
    return { v2ray: txt, clash: yaml };
  },

  async openrunner() {
    const root = await fetchText('https://freenode.openrunner.net/');
    const links = parseLinks(root, 'a.mb-5');
    const page = await fetchText(links[0]);
    const codes = (page.match(/<code[^>]*>([^<]+)<\/code>/g) || []).map(c => c.replace(/<[^>]+>/g, ''));
    const txt = codes.find(c => c.trim().endsWith('.txt')) || '';
    const yaml = codes.find(c => c.trim().endsWith('.yaml')) || '';
    return { v2ray: txt, clash: yaml };
  },

  async clashnode() {
    const root = await fetchText('https://clashnode.cc');
    const links = parseLinks(root, 'main a');
    const page = await fetchText(links[0]);
    const ps = (page.match(/<p[^>]*>([^<]+)<\/p>/g) || []).map(p => p.replace(/<[^>]+>/g, ''));
    return {
      v2ray: ps.find(t => t.endsWith('.txt')) || '',
      clash: ps.find(t => t.endsWith('.yaml')) || '',
      singbox: ps.find(t => t.endsWith('.json')) || '',
    };
  },
};

// ==========  路由表  ==========
const ROUTER = {
  '/c/:site':  'clash',
  '/v/:site':  'v2ray',
  '/s/:site':  'singbox',
  '/m/:site':  'mihomo',
  '/b/changfeng': 'base64',
};

// ==========  主入口  ==========
addEventListener('fetch', event => {
  event.respondWith(handle(event.request));
});

async function handle(req) {
  const url = new URL(req.url);
  const path = url.pathname;

  // 首页用法
  if (path === '/') {
    const host = url.origin;
    const html = `<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>免费订阅中转</title></head>
<body>
<h1>用法列表（直接点击或复制到客户端）</h1>
<h3>米贝</h3>
<ul><li><a href="${host}/c/mibei" target="_blank">Clash</a></li>
    <li><a href="${host}/v/mibei" target="_blank">V2Ray</a></li></ul>
<h3>长风</h3>
<ul><li><a href="${host}/c/changfeng" target="_blank">Clash</a></li>
    <li><a href="${host}/v/changfeng" target="_blank">V2Ray</a></li>
    <li><a href="${host}/s/changfeng" target="_blank">SingBox</a></li>
    <li><a href="${host}/m/changfeng" target="_blank">Mihomo</a></li>
    <li><a href="${host}/b/changfeng" target="_blank">Base64</a></li></ul>
<h3>NodeFree</h3>
<ul><li><a href="${host}/c/nodefree" target="_blank">Clash</a></li>
    <li><a href="${host}/v/nodefree" target="_blank">V2Ray</a></li></ul>
<h3>V2rayShare</h3>
<ul><li><a href="${host}/c/v2rayshare" target="_blank">Clash</a></li>
    <li><a href="${host}/v/v2rayshare" target="_blank">V2Ray</a></li></ul>
<h3>OpenRunner</h3>
<ul><li><a href="${host}/c/openrunner" target="_blank">Clash</a></li>
    <li><a href="${host}/v/openrunner" target="_blank">V2Ray</a></li></ul>
<h3>ClashNode</h3>
<ul><li><a href="${host}/c/clashnode" target="_blank">Clash</a></li>
    <li><a href="${host}/v/clashnode" target="_blank">V2Ray</a></li>
    <li><a href="${host}/s/clashnode" target="_blank">SingBox</a></li></ul>
</body></html>`;
    return new Response(html, { headers: { 'content-type': 'text/html;charset=utf-8' } });
  }

  // 匹配路由
  for (const [pattern, type] of Object.entries(ROUTER)) {
    const reg = new RegExp(`^${pattern.replace(':site', '([\\w-]+)')}$`);
    const m = path.match(reg);
    if (!m) continue;
    const site = m[1];
    if (!SOURCES[site]) return notFound();

    try {
      const data = await SOURCES[site]();
      let url = '';
      if (type === 'base64') url = data.base64;
      else url = data[type];
      if (!url) return notFound();
      // 直接 302 到真实订阅地址（减少 Workers 流量）
      return Response.redirect(url, 302);
    } catch (e) {
      return new Response(`后端抓取失败: ${e.message}`, { status: 502 });
    }
  }

  return notFound();
}

function notFound() {
  return new Response('404 Not Found', { status: 404 });
}
